libbeat/outputs/tidb_slow_query/client.go

```diff
@@ -0,0 +1,233 @@
+	if err != nil {
+		return nil, err
+	}
+	c := &client{
+		observer:  observer,
+		timeout:   timeout,
+		database:  database,
+		db:        db,
+		dsn:       dsn,
+		retention: retention,
+		rollStep:  rollStep,
+		stmtCache: cache,
+		log:       logp.NewLogger("tidb_slow_query"),
+	}
+	return c, nil
+}
+
+// Connect try to create a new connection and replace the conn field
+func (c *client) Connect() error {
+	ctx, cancel := context.WithTimeout(context.Background(), c.timeout)
+	defer cancel()
+	if c.conn != nil && c.conn.PingContext(ctx) == nil {
+		return nil
+	}
+	conn, err := c.db.Conn(ctx)
+	if err != nil {
+		return err
+	}
+	c.conn = conn
+	return nil
+}
+
+func (c *client) Close() error {
+	return c.conn.Close()
+}
+
+func (c *client) Publish(ct context.Context, batch publisher.Batch) error {
+	ctx, cancel := context.WithTimeout(ct, c.timeout)
+	defer cancel()
+
+	// todo: support multi events from different cluster (with different table name)
+	event := batch.Events()[0]
+
+	table, err := c.getClusterIDAsTableName(event)
+	if err != nil {
+		c.observer.Dropped(1)
+		batch.Cancelled()
+		return err
+	}
+
+	// get driver statement
+	if !c.stmtCache.Contains(table) {
+		sqlString := insertStmt(c.database, table)
+		s, err := c.conn.PrepareContext(ctx, sqlString)
+		if err != nil {
+			newErr := c.handleInsertError(err, ctx, table, event)
+			batch.RetryEvents(batch.Events())
+			return newErr
+		}
+		c.stmtCache.Add(table, s)
+	}
+	sqlStmt, _ := c.stmtCache.Get(table)
+
+	// get placeholder arguments
+	fields := getFields(event)
+
+	// execute statement
+	_, executionErr := sqlStmt.(*sql.Stmt).ExecContext(ctx, fields...)
+	if executionErr != nil {
+		newErr := c.handleInsertError(executionErr, ctx, table, event)
+		batch.RetryEvents(batch.Events())
+		return newErr
+	}
+
+	c.observer.NewBatch(1)
+	batch.ACK()
+	return nil
+}
+
+// handle no-table-or-partition error -- create them
+// re-throw other unexpected error
+func (c *client) handleInsertError(err error, ctx context.Context, table string, event publisher.Event) error {
+	if err == nil {
+		return nil
+	}
+
+	// delete corrupted cached stmt
+	c.stmtCache.Remove(table)
+
+	// connection will corrupt when error occurs
+	// try to reconnect
+	if err := c.Connect(); err != nil {
+		return err
+	}
+
+	mysqlErr, ok := err.(*mysql.MySQLError)
+	if !ok {
+		return err
+	}
+
+	// create table/partition, could wait for a while
+	switch mysqlErr.Number {
+	case noPartition:
+		return c.createPartitions(ctx, table, event.Content.Timestamp)
+	case noTable:
+		return c.createTable(ctx, table, event.Content.Timestamp)
+	default:
+		return mysqlErr
+	}
+}
+
+func getFields(event publisher.Event) []interface{} {
+	r := make([]interface{}, 0, len(orderedColumn))
+	for _, c := range orderedColumn {
+		// expect nil if fields not exist
+		v, _ := event.Content.GetValue(c)
+		r = append(r, v)
+	}
+	return r
+}
+
+func (c *client) getClusterIDAsTableName(event publisher.Event) (string, error) {
```

Comment:
Recommend to provide two functions:

1. `extractClusterID(event)`
2. `buildTableName(clusterID)`